### PR TITLE
HIVE-25734: Wrongly-typed constant in case expression leads to incorrect empty result

### PR DESCRIPTION
**Pull request by asolimando**
### What changes were proposed in this pull request?

1) `HivePointLookupOptimizerRule` should not attempt at merging `IN` clauses when the constant types is different, we fix this in the current PR.

2) Another issue emerging in the added qtest query is that constants in `case when` expressions are not typed following the type of the predicate they will be compared too. In the qtest this leads to the creation of `IN` clauses with constants of type `INTEGER` where `SMALLINT` type would have been more appropriate, which makes the optimizer missing opportunities.

### Why are the changes needed?

1) Merging `IN` clauses with incompatible types can lead to empty results while some are expected.

2) The second issue does not harm correctness but limits rewrite opportunities.

### Does this PR introduce _any_ user-facing change?

Queries involving `case when` might have better typing for constants and this can lead to further optimizations, but no user-facing changes apart from correcting the bug.

### How was this patch tested?

The first issue is covered by additional unit tests in `TestHivePointLookupOptimizerRule`, the second one by a new qtest and a new unit test class `TestRexNodeConverter`.

### CODE REVIEW

File: ql/src/test/queries/clientpositive/cbo_case_when_wrong_type.q [hunk 1]
```diff
@@ -0,0 +1,20 @@
+create external table test_case (row_seq smallint, row_desc string) stored as parquet;
```
Comment (kgyrtkirk) [inline ql/src/test/queries/clientpositive/cbo_case_when_wrong_type.q:1]: can we also add a simpler testcase like:
```
explain select 1 from t where a in (1,2,3) and case a when 1S then true when 2S then true end;
```

**Review by kgyrtkirk — APPROVED** [review]

File: ql/src/test/queries/clientpositive/cbo_case_when_wrong_type.q (hunk 1 above)
Comment (asolimando) [reply ql/src/test/queries/clientpositive/cbo_case_when_wrong_type.q:1]: Sure, would you prefer a separate qtest file or to integrate it into this one?

**Comment by asolimando** [comment]
There were 2 test failures:
- testCliDriver[query39] – org.apache.hadoop.hive.cli.TestTezTPCDS30TBPerfCliDriver (updated but the change looks good, it's exactly our use case)
- testCliDriver[udf_unix_timestamp] – org.apache.hadoop.hive.cli.TestBeeLineDriver (unrelated, passes locally, it proved flacky for me in the past this test)

On top of the updated `.q.output` for `query39` I have added `TestRexNodeConverter` class for unit tests (I have added some corner cases with null/nullability and all seems in order.

I had first expanded the `qtest` file with the second variant of `case when` expression (`case when a = 1` instead of `case a when 1`, where the former was already working as expected), but I have finally simplified the test as you suggested.

If the tests pass and you don't have more comments I will squash the commits and upgrade the commit message.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HivePointLookupOptimizerRule.java [hunk 2]
```diff
@@ -655,6 +665,26 @@ private static RexNode handleAND(RexBuilder rexBuilder, RexCall call) {
...
 
+    /**
+     * Check if the type of nodes in the two collections is homogeneous within the collections
+     * and identical between them.
+     * @param nodes1 the first collection of nodes
+     * @param nodes2 the second collection of nodes
+     * @return true if nodes in both collections is unique and identical, false otherwise
+     */
+    private static boolean shareSameType(Collection<RexNode> nodes1, Collection<RexNode> nodes2) {
+      Set<SqlTypeName> types1 = nodes1.stream()
```
[the remark this answers is not in the document]
Comment (ayushtkn) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HivePointLookupOptimizerRule.java:676]: Can any one of this work-

```
     return Stream.of(nodes1, nodes2).flatMap(Collection::stream).map(n -> n.getType().getSqlTypeName()).distinct()
       .count() == 1;
```

```
   return CollectionUtils.union(nodes1, nodes2).stream().map(n -> n.getType().getSqlTypeName()).distinct().count()
          == 1;
```

```     
 return Stream.concat(nodes1.stream(),nodes2.stream()).map(n-> n.getType().getSqlTypeName()).distinct().count() == 1;
```

[the remark this answers is not in the document]
Comment (asolimando) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HivePointLookupOptimizerRule.java:676]: Thanks @ayushtkn, the first one is really what I was looking for, I will adopt it!

**Comment by asolimando** [comment]
Test failures seem totally unrelated but not flaky, those 11 tests failed in the two last runs, how do you usually handle such cases?

**Comment by kgyrtkirk** [comment]
strange...iirc I've seen this test fail randomly before - I've restarted tests on this PR and launched http://ci.hive.apache.org/job/hive-flaky-check/480/ to check `TestStreamingDynamicPartitioning`

**Comment by kgyrtkirk** [comment]
note: if the PR name starts with the associated jira issue id - then a link is added to the jira

**Comment by asolimando** [comment]
Tests passed in the last run, can I squash the commits and rephrase the commit message?

**Comment by kgyrtkirk** [comment]
no need; we will squash it before merging the changes - if you force-push the branch all the earlier comment will loose context; and the review system also looses the marking of which files are marked as viewed
